doc/pandas-2.0/source/internal-architecture.rst

```diff
@@ -0,0 +1,714 @@
+* Bitmaps can propagate their nulls in C through bitwise ``&`` or ``|``
+  operations, which are inexpensive.
+* Getting and setting bits on modern hardware is CPU-inexpensive. For
+  single-pass array operations (like groupbys) on large arrays this may also
+  result in better CPU cache utilization (fewer main-memory reads of the
+  bitmap).
+* Hardware and SIMD "popcount" intrinsics (which can operate on 64-128 bits at
+  a time) can be used to count bits and skip null-handling on segments of data
+  containing no nulls.
+
+Notably, this is the way that PostgreSQL handles null values. For example, we
+might have:
+
+.. code-block::
+
+   [0, 1, 2, NA, NA, 5, 6, NA]
+
+        i: 7 6 5 4 3 2 1 0
+   bitmap: 0 1 1 0 0 1 1 1
+
+Here, the convention of 1 for "not null" (a la PostgreSQL) and
+least-significant bit ordering (LSB "bit endianness") is being used.
+
+Under the new regime, users could simply write:
+
+.. code-block:: python
+
+   s[2] = pandas.NA
+
+and the data type would be unmodified. It may be necessary to write something
+akin to:
+
+.. code-block:: python
+
+   s.to_numpy(dtype=np.float64, na_rep=np.nan)
+
+and that would emulate the current behavior. Attempts to use ``__array__` (for
+example: calling ``np.sqrt`` on the data) would result in an error since we
+will likely want to refuse to make a guess as for what casting behavior the
+user desires.
+
+Tradeoffs
+~~~~~~~~~
+
+One potential downside of the bitmap approach is that missing data implemented
+outside of NumPy's domain will need to be explicitly converted if it is needed
+in another library that only knows about NumPy. I argue that this is better
+than the current implicit conversion which could yield data loss (for integers
+falling outside the exact representable range for ``float64``).
+
+Removal of BlockManager / new DataFrame internals
+=================================================
+
+Deep inside the belly pandas objects, there is a data structure called
+``BlockManager`` which, at a high level, is responsible for managing the
+physical arrays where the data inside a Series or DataFrame is looked
+after (also Panel / PanelND structure, even though these are on their way to
+deprecation).
+
+While this data structure has served pandas well since its birth 5 years ago
+(Summer 2011), it has a number of problems that make its removal and
+replacement with something else an attractive option.
+
+The goal of this section is to explain what the BlockManager is, why it exists
+at all, and why we should consider removing it.
+
+What is ``BlockManager`` and why does it exist?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The reason that ``BlockManager`` exists at all goes back to some ancient pandas
+history. Originally, the data in ``pandas.DataFrame`` was stored in a Python
+``dict`` object. If you pull up pandas 0.1 or 0.2, you will see this.
+
+Since the business logic of pandas's internals was originally implemented in
```

Comment:
Thanks for creating that, I will link to in the document!

I will also definitely mention here that zero-copy creation from 2D ndarrays was a primary motivation (I had dealt with a number of applications in the past where this was a bottleneck). I suspect we'll be able to make the constructor significantly faster in libpandas -- maybe with some tricks, like processing L2-ish cache size contiguous chunks of the matrix at a time to reduce the perf hit of strided memory access.